staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -315,23 +320,96 @@ func strategicPatchObject(
 	if err != nil {
 		return nil, err
 	}
-	if err := runtime.DecodeInto(p.codec, p.patchBytes, versionedObjToUpdate); err != nil {
-		return nil, err
+	return runtime.DefaultUnstructuredConverter.ToUnstructured(vo)
+}
+
+func (p *applyPatcher) LastAppliedObject(obj runtime.Object, workflow string) (map[string]interface{}, error) {
+	acsor, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get accessor: %v", err)
 	}
-	// For now, instead of performing a structured merge we will just attempt a PUT
-	// So we just convert the object back to unversioned (aka internal version).
-	unversionedObjToUpdate, err := p.toUnversioned(versionedObjToUpdate)
+	last := make(map[string]interface{})
+	if acsor.GetLastApplied()[workflow] != "" {
+		if err := json.Unmarshal([]byte(acsor.GetLastApplied()[workflow]), &last); err != nil {
+			return nil, fmt.Errorf("couldn't unmarshal last applied field: %v", err)
+		}
+	}
+	return last, nil
+}
+
+func (p *applyPatcher) PatchObject() (map[string]interface{}, error) {
+	patch := make(map[string]interface{})
+	if err := yaml.Unmarshal(p.patchBytes, &patch); err != nil {
+		return nil, fmt.Errorf("couldn't unmarshal patch object: %v (patch: %v)", err, string(p.patchBytes))
+	}
+	return patch, nil
+}
+
+func (p *applyPatcher) OutputObject(result interface{}) (runtime.Object, error) {
+	voutput, err := p.creater.New(p.kind)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to create empty output object: %v", err)
 	}
-	return unversionedObjToUpdate, nil
+
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(result.(map[string]interface{}), voutput)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert merge result back: %v", err)
+	}
+	p.defaulter.Default(voutput)
+
+	uoutput, err := p.toUnversioned(voutput)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert to unversioned: %v", err)
+	}
+
+	return uoutput, nil
+}
+
+func (p *applyPatcher) applyPatchToCurrentObject(currentObject runtime.Object) (runtime.Object, error) {
+	live, err := p.LiveObject(currentObject)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get live object: %v", err)
+	}
+
+	last, err := p.LastAppliedObject(currentObject, workflowId)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get last applied object: %v", err)
+	}
+	patch, err := p.PatchObject()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get patch object: %v", err)
+	}
+
+	elmt, err := parse.CreateElement(last, patch, live, p.model)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse elements: %v", err)
+	}
+	result, err := elmt.Merge(strategy.Create(strategy.Options{}))
+	if err != nil {
+		return nil, fmt.Errorf("failed to merge elements: %v", err)
+	}
+
+	output, err := p.OutputObject(result.MergedResult)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert merge result: %v", err)
+	}
+
+	// TODO(apelisse): Update last applied
+
```

Comment:
TODO: check for conflicts with other workflows

Reply:
Are we planning on having this for alpha?

Reply:
no, but I definitely don't want to forget